File: app/src/main/java/edu/sunypoly/inventoryapp/Authenticator.java
package edu.sunypoly.inventoryapp;

import android.os.AsyncTask;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.concurrent.ExecutionException;

public class Authenticator {
    private final String TAG = this.getClass().getSimpleName();
    //characters reserved by JSON and SQL standards
     static final char[] ILLEGAL_CHARACTERS = new char[]{'{', '}', '[', ']', '/', '\\', ':', '#',
            ',', '?', '&', '=', '<', '>', '(', ')', '*', '^', '!', '~', '-', '|', ';', '%'};

    private static final Authenticator authenticator = new Authenticator();

    private boolean loggedIn;
    private String userKey;
    private static URL url;

    private Authenticator() {
        try {
            url = new URL("http://150.156.202.112:8000/inventory");
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
    }

    public static Authenticator getInstance() {
        return authenticator;
    }


    boolean login(String username, String password) {
        //TODO
        loggedIn = true;
        return loggedIn;
    }

    void logout() {
        loggedIn = false;
    }

    boolean isLoggedIn(){
        return loggedIn;
    }

    boolean addItem(InventoryItem item) {
        if (loggedIn) {
            AddItemsTask addItemsTask = new AddItemsTask(item);
            addItemsTask.execute(this);
            try {
                return addItemsTask.get();
            } catch (ExecutionException e) {
                e.printStackTrace();
                Log.e(TAG, "ExecutionException " + e.getMessage());
                return false;
            } catch (InterruptedException e) {
                e.printStackTrace();
                Log.e(TAG, "InterruptedException " + e.getMessage());
                return false;
            }
        } else {
            Log.e(TAG, "Not logged in.");
            return false;
        }
    }

    private boolean addItemToDatabase(InventoryItem item) {
        URL url;
        try {
            url = new URL("http://150.156.202.112:8000/inventory");
            HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
            httpURLConnection.setRequestMethod("POST");

            Gson gson = new GsonBuilder().create();
            String json = gson.toJson(item);

            Log.v(TAG, json);


            httpURLConnection.setDoOutput(true);
            DataOutputStream out = new DataOutputStream(httpURLConnection.getOutputStream());
            out.writeBytes(json);
            out.flush();
            out.close();

            httpURLConnection.setConnectTimeout(5000);
            httpURLConnection.setReadTimeout(5000);

            String input = inputStreamToString(httpURLConnection.getInputStream());
            Log.v(TAG, input);

            return true;
        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return false;
    }

    private static class AddItemsTask extends AsyncTask<Authenticator, Void, Boolean> {

        private InventoryItem item;

        AddItemsTask(InventoryItem item) {
            this.item = item;
        }

        @Override
        protected Boolean doInBackground(Authenticator... authenticators) {
            return authenticators[0].addItemToDatabase(item);
        }
    }

    ArrayList<InventoryItem> getItems() {
        if (loggedIn) {
            ReadFilesTask readFilesTask = new ReadFilesTask();
            readFilesTask.execute(this);
            try {
                return readFilesTask.get();
            } catch (ExecutionException e) {
                e.printStackTrace();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        } else {
            Log.e(TAG, "You must be logged in to perform that action.");
        }
        return null;
    }

    boolean deleteItem(InventoryItem item) {
        if (loggedIn) {
            DeleteItemsTask deleteItemsTask = new DeleteItemsTask();
            deleteItemsTask.execute(item);
            try {
                return deleteItemsTask.get();
            } catch (ExecutionException e) {
                e.printStackTrace();
                Log.e(TAG, "ExecutionException " + e.getMessage());
                return false;
            } catch (InterruptedException e) {
                e.printStackTrace();
                Log.e(TAG, "InterruptedException " + e.getMessage());
                return false;
            }
        } else {
            Log.e(TAG, "Not logged in.");
            return false;
        }
    }

    private static class DeleteItemsTask extends AsyncTask<InventoryItem, Void, Boolean> {
        private final String TAG = "DeleteItemsTask";

        @Override
        protected Boolean doInBackground(InventoryItem... items) {
            try {

                HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
                httpURLConnection.setRequestMethod("DELETE");

                Gson gson = new GsonBuilder().create();
                String json = gson.toJson(items[0]);

                Log.v(TAG, json);


                httpURLConnection.setDoOutput(true);
                DataOutputStream out = new DataOutputStream(httpURLConnection.getOutputStream());
                out.writeBytes(json);
                out.flush();
                out.close();

                httpURLConnection.setConnectTimeout(5000);
                httpURLConnection.setReadTimeout(5000);

                String input = inputStreamToString(httpURLConnection.getInputStream());
                Log.v(TAG, input);

                return true;
            } catch (MalformedURLException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            }

            return false;
        }
    }

    private ArrayList<InventoryItem> getItemsFromServer() {
        URL url;
        try {
            url = new URL("http://150.156.202.112:8000/inventory");
            HttpURLConnection httpCon = (HttpURLConnection) url.openConnection();
            httpCon.setRequestMethod("GET");

            Gson gson = new GsonBuilder().create();

            String input = inputStreamToString(httpCon.getInputStream());

            InventoryItem[] inventoryItems = gson.fromJson(input, InventoryItem[].class);

            if (inventoryItems != null) {
                return new ArrayList<>(Arrays.asList(inventoryItems));
            } else {
                return null;
            }
        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static class ReadFilesTask extends AsyncTask<Authenticator, Void, ArrayList<InventoryItem>> {

        @Override
        protected ArrayList<InventoryItem> doInBackground(Authenticator... authenticators) {
            return authenticators[0].getItemsFromServer();
        }
    }


     static String inputStreamToString(InputStream is) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(is));
        StringBuilder result = new StringBuilder();
        String line;
        while ((line = br.readLine()) != null) {
            result.append(line);
        }
        return result.toString();
    }


}

File: app/src/main/java/edu/sunypoly/inventoryapp/Authenticator.kt
package edu.sunypoly.inventoryapp

import android.os.AsyncTask
import android.util.Log
import com.google.gson.GsonBuilder
import java.io.*
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.*
import java.util.concurrent.ExecutionException

/**
 * Handles all the server stuff
 */
class Authenticator//Private constructor is used to make the class a singleton
private constructor() {
    //Used for nicer Log calls
    private val TAG = this.javaClass.simpleName

    /**
     * @return true if user is logged in, else false
     */
    internal var isLoggedIn: Boolean = false

    /**
     * Returns list of all items in a AuthenticatorStatus.ListStatus if the server is running,
     * else returns an appropriate error status
     * @return
     */
    internal val items: AuthenticatorStatus
        get() {
            if (isLoggedIn) {
                val readFilesTask = ReadFilesTask()
                readFilesTask.execute()
                try {
                    return readFilesTask.get()
                } catch (e: Exception) {
                    return AuthenticatorStatus.ServerError
                }

            } else {
                return AuthenticatorStatus.AuthError
            }
        }


    init {
        try {
            url = URL("http://150.156.202.112:8000/inventory")
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        }

    }


    /**
     * Handles logging in
     * Obviously missing some things
     * @param username
     * @param password
     * @return
     */
    internal fun login(username: String, password: String): Boolean {
        //TODO
        isLoggedIn = true
        return isLoggedIn
    }

    /**
     * Logs the user out.
     */
    internal fun logout() {
        isLoggedIn = false
    }

    /**
     * Starts async task to add the item
     * @param item
     * @return
     */
    internal fun addItem(item: InventoryItem): AuthenticatorStatus {
        return if (isLoggedIn) {
            val addItemsTask = AddItemsTask(item)
            addItemsTask.execute()
            try {
                addItemsTask.get()
            } catch (e: Exception) {
                AuthenticatorStatus.ServerError
            }

        } else {
            Log.e(TAG, "Not logged in.")
            AuthenticatorStatus.AuthError
        }
    }


    /**
     * Asynchronously adds the item
     * Probably helpful to look at the documentation for android.os.AsyncTask
     */
    private class AddItemsTask internal constructor(private val item: InventoryItem) : AsyncTask<Void, Void, AuthenticatorStatus>() {


        override fun doInBackground(vararg voids: Void): AuthenticatorStatus {
            return addItemToDatabase(item)
        }

        private fun addItemToDatabase(item: InventoryItem): AuthenticatorStatus {
            val url: URL
            try {
                //Defines URL
                url = URL("http://150.156.202.112:8000/inventory")
                //Tries to connect to URL via HTTP protocol
                val httpURLConnection = url.openConnection() as HttpURLConnection
                httpURLConnection.requestMethod = "POST"

                //Inits GSON
                val gson = GsonBuilder().create()
                //GSON converts item to json
                val json = gson.toJson(item)

                httpURLConnection.doOutput = true
                //Connects to output stream
                val out = DataOutputStream(httpURLConnection.outputStream)
                out.writeBytes(json)//writes the json
                out.flush()//actually writes the json (Java stream thing, this needs to be called)
                out.close()//Closes the stream

                //Not necessary calls, but helpful if you want to fine tune things for some reason
                httpURLConnection.connectTimeout = 5000
                httpURLConnection.readTimeout = 5000

                val input = inputStreamToString(httpURLConnection.inputStream)

                return AuthenticatorStatus.AddedItem
            } catch (e: MalformedURLException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }

            return AuthenticatorStatus.ServerError
        }

    }

    internal fun deleteItem(item: InventoryItem): Boolean {
        if (isLoggedIn) {
            val deleteItemsTask = DeleteItemsTask()
            deleteItemsTask.execute(item)
            try {
                return deleteItemsTask.get()
            } catch (e: ExecutionException) {
                e.printStackTrace()
                Log.e(TAG, "ExecutionException " + e.message)
                return false
            } catch (e: InterruptedException) {
                e.printStackTrace()
                Log.e(TAG, "InterruptedException " + e.message)
                return false
            }

        } else {
            Log.e(TAG, "Not logged in.")
            return false
        }
    }

    private class DeleteItemsTask : AsyncTask<InventoryItem, Void, Boolean>() {
        private val TAG = "DeleteItemsTask"

        override fun doInBackground(vararg items: InventoryItem): Boolean? {
            try {

                val httpURLConnection = url!!.openConnection() as HttpURLConnection
                httpURLConnection.requestMethod = "DELETE"

                val gson = GsonBuilder().create()
                val json = gson.toJson(items[0])

                Log.v(TAG, json)


                httpURLConnection.doOutput = true
                val out = DataOutputStream(httpURLConnection.outputStream)
                out.writeBytes(json)
                out.flush()
                out.close()

                httpURLConnection.connectTimeout = 5000
                httpURLConnection.readTimeout = 5000

                val input = inputStreamToString(httpURLConnection.inputStream)
                Log.v(TAG, input)

                return true
            } catch (e: MalformedURLException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }

            return false
        }
    }


    class ReadFilesTask : AsyncTask<Void, Void, AuthenticatorStatus>() {

        private val itemsFromServer: AuthenticatorStatus
            get() {
                val url: URL
                try {
                    url = URL("http://150.156.202.112:8000/inventory")
                    val httpCon = url.openConnection() as HttpURLConnection
                    httpCon.requestMethod = "GET"

                    val gson = GsonBuilder().create()

                    val input = inputStreamToString(httpCon.inputStream)

                    val inventoryItems = gson.fromJson(input, Array<InventoryItem>::class.java)

                    return if (inventoryItems != null) {
                        AuthenticatorStatus.ListStatus("Successfully got items.", ArrayList(Arrays.asList(*inventoryItems)))
                    } else {
                        AuthenticatorStatus.NoItems
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                    return AuthenticatorStatus.ServerError
                }

            }

        override fun doInBackground(vararg voids: Void): AuthenticatorStatus {
            return itemsFromServer
        }
    }

    companion object {
        //characters reserved by JSON and SQL standards
        internal val ILLEGAL_CHARACTERS = charArrayOf('{', '}', '[', ']', '/', '\\', ':', '#', ',', '?', '&', '=', '<', '>', '(', ')', '*', '^', '!', '~', '-', '|', ';', '%')

        //This is used to make the class a singleton
        //This is how other classes access this class
        val instance = Authenticator()
        private var url: URL? = null


        @Throws(IOException::class)
        internal fun inputStreamToString(`is`: InputStream): String {
            val br = BufferedReader(InputStreamReader(`is`))
            val result = StringBuilder()
            var line = br.readLine()
            while (line != null) {
                result.append(line)
                line = br.readLine()
            }
            return result.toString()
        }
    }


}
